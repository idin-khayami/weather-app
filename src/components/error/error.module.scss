@import '../../styles/vars';

.error {
  padding: 1rem;
  border: 2px solid $dangerColor;
  background-color: rgba($dangerColor, 0.4);
  border-radius: 10px;

  p {
    font-size: 1.1rem;
    font-weight: 500;
  }
}
