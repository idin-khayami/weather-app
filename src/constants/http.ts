export const apiConfigURL =
  process.env.NEXT_PUBLIC_BASE_API_URL ||
  'https://api.openweathermap.org/data/2.5';

export const apiConfigStaleTime = 4 * 60 * 60 * 1000; // 4h

export const apiConfigKey = process.env.NEXT_PUBLIC_API_KEY as string;

export const HTTP_STATUS_CODES = {
  BAD_REQUEST: 400,
  UNAUTHORIZED: 401,
  FORBIDDEN: 403,
  NOT_FOUND: 404,
  INTERNAL_SERVER_ERROR: 500,
};
