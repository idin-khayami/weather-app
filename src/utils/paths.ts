export interface Paths {
  homepage: string;
  getModifyRoute: (id: string) => string;
}

export const paths: Paths = {
  homepage: '/',
  getModifyRoute: (id: string) => `/modify/${id}`,
};
