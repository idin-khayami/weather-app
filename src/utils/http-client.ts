import { apiConfigURL, apiConfigKey } from '@/constants/http';

export const httpClient = {
  async get<T = unknown>(
    url: string,
    query: Record<string, string>,
  ): Promise<T> {
    const queryStr = new URLSearchParams(
      Object.keys(query).length > 0
        ? { ...query, apikey: apiConfigKey }
        : { apikey: apiConfigKey },
    ).toString();

    const res = await fetch(`${apiConfigURL}${url}?${queryStr}`);

    return await res.json();
  },
};
