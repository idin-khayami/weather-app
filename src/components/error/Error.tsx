import clsx from 'clsx';

import { ComponentBaseProps } from '@/types/global.types';
import { DEFAULT_ERROR_MESSAGE } from '@/utils/constants';

import styles from './error.module.scss';

interface ErrorProps extends ComponentBaseProps {
  hasError?: boolean;
  errorMessage?: string;
}

export const Error = ({
  dataTestId,
  errorMessage = DEFAULT_ERROR_MESSAGE,
  hasError = true,
  className,
}: ErrorProps) => {
  if (!hasError) return null;

  return (
    <div data-testid={dataTestId} className={clsx([styles.error, className])}>
      <p>{errorMessage}</p>
    </div>
  );
};
