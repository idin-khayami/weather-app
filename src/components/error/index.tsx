import { Error } from '@/components/error/Error';

export { Error };
