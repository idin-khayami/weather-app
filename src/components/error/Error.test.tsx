import { render, screen } from '@testing-library/react';

import { Error } from '@/components/error/Error';
import { DEFAULT_ERROR_MESSAGE } from '@/utils/constants';

describe('Component: error component', () => {
  it('should render error by default', () => {
    render(<Error dataTestId="testingError" />);

    expect(screen.getByTestId('testingError')).toBeInTheDocument();
    expect(screen.getByTestId('testingError')).toMatchSnapshot();
  });

  it('should not have error if hasError is false', () => {
    render(<Error dataTestId="testingError" hasError={false} />);

    expect(screen.queryByTestId('testingError')).toBeNull();
  });

  it('should be able to render default error message', () => {
    render(<Error dataTestId="testingError" hasError={true} />);

    expect(screen.getByTestId('testingError')).toHaveTextContent(
      DEFAULT_ERROR_MESSAGE,
    );
  });

  it('should be able to render specified error message', () => {
    render(
      <Error
        dataTestId="testingError"
        hasError={true}
        errorMessage="Awesome Error"
      />,
    );

    expect(screen.getByTestId('testingError')).toHaveTextContent(
      'Awesome Error',
    );
  });
});
