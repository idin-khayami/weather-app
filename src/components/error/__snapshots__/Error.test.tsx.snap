// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: error component should render error by default 1`] = `
<div
  class="error"
  data-testid="testingError"
>
  <p>
    Something went wrong!, please try again.
  </p>
</div>
`;
